extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,580 @@
+   * modify the change entry to be correct.
+   * @param {!Object} changes
+   * @return {!Object} Corrected intersection change entry.
+   */
+  correctChanges(changes) {
+    const iframeRect = this.iframe_ ? this.iframe_.getBoundingClientRect() :
+      this.creativeSize_;
+    const ampAdRect = this.baseInstance_.element.getBoundingClientRect();
+    const widthCorrection = (ampAdRect.width - iframeRect.width) / 2;
+    const heightCorrection = (ampAdRect.height - iframeRect.height) / 2;
+    changes.boundingClientRect.right -= widthCorrection;
+    changes.boundingClientRect.top += heightCorrection;
+    changes.boundingClientRect.bottom -= heightCorrection;
+    changes.boundingClientRect.left += widthCorrection;
+    return changes;
+  }
+
+  /**
+   * Converts an IntersectionObserver-formatted change message to the
+   * geometry update format expected by GPT Safeframe.
+   * @param {!Object} changes IntersectionObserver formatted
+   *   changes.
+   * @return {string} Safeframe formatted changes.
+   * @private
+   */
+  formatGeom_(changes) {
+    const percInView = (a1, b1, a2, b2) => {
+      const lengthInView = (b2 >= b1) ? b1 - a2 : b2;
+      const percInView = lengthInView / (b2 - a2);
+      if (percInView < 0) {
+        return 0;
+      } else if (percInView > 1) {
+        return 1;
+      } else {
+        return percInView;
+      }
+    };
+    changes = this.correctChanges(changes);
+    const expandBounds = this.getExpandBounds(changes);
+    this.currentGeometry_ = {
+      'windowCoords_t': changes.rootBounds.top,
+      'windowCoords_r': changes.rootBounds.right,
+      'windowCoords_b': changes.rootBounds.bottom,
+      'windowCoords_l': changes.rootBounds.left,
+      'frameCoords_t': changes.boundingClientRect.top,
+      'frameCoords_r': changes.boundingClientRect.right,
+      'frameCoords_b': changes.boundingClientRect.bottom,
+      'frameCoords_l': changes.boundingClientRect.left,
+      'styleZIndex': this.baseInstance_.element.style.zIndex,
+      'allowedExpansion_t': expandBounds.height,
+      'allowedExpansion_r': expandBounds.width,
+      'allowedExpansion_b': expandBounds.height,
+      'allowedExpansion_l': expandBounds.width,
+      'yInView': percInView(changes.rootBounds.top,
+          changes.rootBounds.bottom,
+          changes.boundingClientRect.top,
+          changes.boundingClientRect.bottom),
+      'xInView': percInView(changes.rootBounds.left,
+          changes.rootBounds.right,
+          changes.boundingClientRect.left,
+          changes.boundingClientRect.right),
+    };
+    return JSON.stringify(this.currentGeometry_);
+  }
+
+  /**
+   * Calculates and returns the allowed expansion values for height and width.
+   * @param {!Object} changes Intersection change entry.
+   * @returns {!Object<string, number>}
+   */
+  getExpandBounds(changes) {
+    const frameHeight = changes.boundingClientRect.bottom -
```

Comment:
Can this just be boundingClientRect.height?  Check on all of these if height/width are present...